WIP: Adds an option for warning on error, to detect all missing columns

GuyPaddock opened the pull request.

conor-pappas committed:
Adds behavior for warning on columns that are missing without hard failing

conor-pappas committed:
Fixing syntax

Commit:
Merge pull request #1 from conor-pappas/soft_failure_support

Adds behavior for warning on columns that are missing without hard fa…

GuyPaddock committed:
[ECOM-1017] Expands on the warning support to also handle the case where the config defines a column that no longer exists.

GuyPaddock committed:
Centralizes error handling, and adds a way for the Obfuscator to report errors back to the caller if it's in `:warn` mode.

GuyPaddock committed:
Fixes handling of errors that are concatenated together into a single string.

GuyPaddock committed:
Fixes two other places where warnings should be routed through the new `handle_error` method.

GuyPaddock committed:
Makes `handle_error` public so it can be called from the remaining places.

cantino commented on lib/my_obfuscate.rb:
```diff
@@ -80,12 +97,15 @@ def missing_column_list(table_name, columns)
-      raise RuntimeError.new(error_message)
+    unless unspecified_columns_behavior == :ignore
+      missing_columns = missing_column_list(table_name, columns)
+      unless missing_columns.length == 0
```
Might be more readable as `unless missing_columns.length == 0 && unspecified_columns_behavior != :ignore`

cantino commented on lib/my_obfuscate.rb:
```diff
@@ -97,14 +117,34 @@ def obfuscate_bulk_insert_line(line, table_name, columns, ignore = nil)
+      # Prevent errors with extra columns
+      if (unspecified_columns_behavior != :fail) && table_config
+        extra_columns = extra_column_list(table_name, columns)
+        table_config.reject! { |k,v| extra_columns.include?(k) }
```
Are you sure you want to mutate `table_config`?

GuyPaddock replied:
Good catch -- I was thinking that I wanted to mutate in-place since I do want to reject the missing columns, but was forgetting that this hash is passed in from outside the method.

I'm switching this up now.

cantino commented:
Thanks @GuyPaddock, this seems like a reasonable idea.  Would you be able to add some specs as well?

GuyPaddock committed:
- Moves error handling to earlier in the call sequence to:
-- Cut down on the number of errors we generate (now we generate only one error per table rather than one per row); AND
-- Allow the script to detect missing columns even for empty tables.
- Addresses PR feedback.

GuyPaddock commented:
Thanks, @cantino. I will add some tests this weekend. Please consider this a WIP for the time being.

GuyPaddock committed:
- Moves the concern of interpreting the error mode to the error handler.
- Renames the error handler, for clarity.
- Makes failure the default way to handle missing columns (for data security).

GuyPaddock committed:
Fixes lazy initialization behavior.

GuyPaddock committed:
Fixes calls to check config from statement parsers.

GuyPaddock committed:
Fixes quoted column names from Postgres, and closes mavenlink#16.

GuyPaddock committed:
Ensures we're matching the start and end of the string when finding quotes.

GuyPaddock committed:
Fixes the way errors are handled so that we don't join messages together only to split them apart later in the error handler.

GuyPaddock committed:
Moves the pruning of missing columns out of `obfuscate_bulk_insert_line`.

GuyPaddock committed:
Renames "unspecified column behavior" to "column mismatch behavior" (3 characters shorter!).

GuyPaddock committed:
Removes unnecessary `attr_accessor` and fixes the mutator for `column_mismatch_behavior`.

GuyPaddock committed:
Fixes bad `split` call for error messages (meant to kill this in ed5d3f4).
